src/main/java/com/mongodb/kafka/connect/source/StartedMongoSourceTask.java

```diff
@@ -210,68 +211,80 @@ private List<SourceRecord> pollInternal() {
-                    publishFullDocumentOnlyTombstoneOnDelete
-                        && !changeStreamDocument.containsKey(FULL_DOCUMENT);
-                if (publishFullDocumentOnly) {
-                  if (changeStreamDocument.containsKey(FULL_DOCUMENT)
-                      && changeStreamDocument.get(FULL_DOCUMENT).isDocument()) {
-                    valueDocument = Optional.of(changeStreamDocument.getDocument(FULL_DOCUMENT));
-                  }
-                } else {
-                  valueDocument = Optional.of(changeStreamDocument);
-                }
-
-                if (valueDocument.isPresent() || isTombstoneEvent) {
-                  BsonDocument valueDoc = valueDocument.orElse(new BsonDocument());
-                  LOGGER.trace("Adding {} to {}: {}", valueDoc, topicName, sourceOffset);
-
-                  if (valueDoc instanceof RawBsonDocument) {
-                    int sizeBytes = ((RawBsonDocument) valueDoc).getByteBuffer().limit();
-                    statisticsManager.currentStatistics().getMongodbBytesRead().sample(sizeBytes);
-                  }
-
-                  BsonDocument keyDocument;
-                  if (sourceConfig.getKeyOutputFormat() == MongoSourceConfig.OutputFormat.SCHEMA) {
-                    keyDocument = changeStreamDocument;
-                  } else if (sourceConfig.getBoolean(DOCUMENT_KEY_AS_KEY_CONFIG)
-                      && changeStreamDocument.containsKey(DOCUMENT_KEY_FIELD)) {
-                    keyDocument = changeStreamDocument.getDocument(DOCUMENT_KEY_FIELD);
-                  } else {
-                    keyDocument = new BsonDocument(ID_FIELD, changeStreamDocument.get(ID_FIELD));
-                  }
-
-                  createSourceRecord(
-                          keySchemaAndValueProducer,
-                          isTombstoneEvent
-                              ? TOMBSTONE_SCHEMA_AND_VALUE_PRODUCER
-                              : valueSchemaAndValueProducer,
-                          sourceOffset,
-                          topicName,
-                          keyDocument,
-                          valueDoc)
-                      .map(sourceRecords::add);
-                }
-              }
-            });
+      // if isCopying is true, we want to set the COPY_KEY flag so that kafka has context that a
+      // copy is in progress. However, for the last document that we are copying, we should not set
+      // this flag because the copy has completed, otherwise we are relying on future change stream
+      // events to signify that we are no longer copying. We also need to set the _id field to be a
+      // valid resume token, which during copying exists in the cachedResumeToken variable.
+      boolean lastDocument = !batchIterator.hasNext();
+      boolean noMoreDataToCopy = copyDataManager != null && !copyDataManager.isCopying();
+      if (isCopying && lastDocument && noMoreDataToCopy) {
+        sourceOffset.put(ID_FIELD, cachedResumeToken.toJson());
```

Comment:
great find! Should we add a test that confirms we're setting the expected resume token?

Reply:
I think we got coverage of this with the recent e2e tests I added around restarting after copying. Let me know if you disagree.